File: Sources/DebridStreamer/Services/Debrid/DebridManager.swift
import Foundation

/// Manages multiple debrid services and routes requests to the best available one.
actor DebridManager {
    private var services: [DebridServiceProtocol] = []
    private let secretStore: any SecretStore

    init(secretStore: any SecretStore = KeychainSecretStore()) {
        self.secretStore = secretStore
    }

    /// Configure services from stored configs.
    func configure(configs: [DebridConfig]) async {
        let activeConfigs = configs
            .filter { $0.isActive }
            .sorted { $0.priority < $1.priority }

        var resolvedServices: [DebridServiceProtocol] = []
        for config in activeConfigs {
            guard let token = await resolveToken(for: config), !token.isEmpty else {
                continue
            }

            switch config.service {
            case .realDebrid:
                resolvedServices.append(RealDebridService(apiToken: token))
            case .allDebrid:
                resolvedServices.append(AllDebridService(apiToken: token))
            case .premiumize:
                resolvedServices.append(PremiumizeService(apiToken: token))
            case .torBox:
                resolvedServices.append(TorBoxService(apiToken: token))
            }
        }

        services = resolvedServices
    }

    private func resolveToken(for config: DebridConfig) async -> String? {
        guard let secretKey = SecretReference.decode(config.apiToken) else {
            return config.apiToken
        }

        do {
            return try await secretStore.getSecret(for: secretKey)
        } catch {
            #if DEBUG
            print("[DebridManager] Failed to resolve secret for \(config.service.rawValue): \(error)")
            #endif
            return nil
        }
    }

    /// Add a single service.
    func addService(_ service: DebridServiceProtocol) {
        services.append(service)
    }

    /// Get all active service types.
    var activeServiceTypes: [DebridServiceType] {
        services.map(\.serviceType)
    }

    var hasServices: Bool {
        !services.isEmpty
    }

    /// Check cache across all active debrid services.
    ///
    /// For each hash the result binds the highest-priority service that actually
    /// reports the hash as CACHED (so `resolveStream` is never routed to a service
    /// that does not have it cached). Hashes that no service has cached still get a
    /// reported status (so callers can sort/display them), bound to the
    /// highest-priority service that returned a status for them. The merge is fully
    /// deterministic regardless of task completion order, driven by each service's
    /// configured priority index.
    func checkCacheAll(hashes: [String]) async throws -> [String: (service: DebridServiceType, status: CacheStatus)] {
        guard !hashes.isEmpty else { return [:] }

        // Query all services concurrently, carrying each service's priority index so
        // the merge can be made order-independent (services is sorted by priority).
        var collected: [(index: Int, serviceType: DebridServiceType, cache: [String: CacheStatus])] = []
        await withTaskGroup(of: (Int, DebridServiceType, [String: CacheStatus]).self) { group in
            for (index, service) in services.enumerated() {
                group.addTask {
                    let cache = (try? await service.checkCache(hashes: hashes)) ?? [:]
                    return (index, service.serviceType, cache)
                }
            }

            for await result in group {
                collected.append((result.0, result.1, result.2))
            }
        }

        // Deterministic merge by priority index. A cached entry always beats a
        // non-cached one; among entries of equal cached-ness the lower index wins.
        var results: [String: (index: Int, service: DebridServiceType, status: CacheStatus)] = [:]
        for entry in collected.sorted(by: { $0.index < $1.index }) {
            for (hash, status) in entry.cache {
                if let existing = results[hash] {
                    let beatsCacheState = status.isCached && !existing.status.isCached
                    let sameCacheStateLowerIndex = status.isCached == existing.status.isCached && entry.index < existing.index
                    if beatsCacheState || sameCacheStateLowerIndex {
                        results[hash] = (entry.index, entry.serviceType, status)
                    }
                } else {
                    results[hash] = (entry.index, entry.serviceType, status)
                }
            }
        }

        return results.mapValues { (service: $0.service, status: $0.status) }
    }

    /// Resolve a torrent hash to a stream URL using the specified (or first available) service.
    func resolveStream(hash: String, preferredService: DebridServiceType? = nil) async throws -> StreamInfo {
        let service: DebridServiceProtocol

        if let preferred = preferredService {
            guard let found = services.first(where: { $0.serviceType == preferred }) else {
                throw DebridError.networkError("Service \(preferred.displayName) not configured")
            }
            service = found
        } else {
            guard let first = services.first else {
                throw DebridError.networkError("No debrid services configured")
            }
            service = first
        }

        // For Real-Debrid, use the smart resolve flow
        if let rdService = service as? RealDebridService {
            return try await resolveWithRealDebrid(rdService, hash: hash)
        }

        // Generic flow for other services
        let torrentId = try await service.addMagnet(hash: hash)
        try await service.selectFiles(torrentId: torrentId, fileIds: [])
        return try await service.getStreamURL(torrentId: torrentId)
    }

    /// Smart Real-Debrid resolve flow:
    /// 1. Check if torrent already exists in user's list
    /// 2. If not, add magnet
    /// 3. Select files if needed
    /// 4. Poll for "downloaded" status
    /// 5. Unrestrict and return stream URL
    private func resolveWithRealDebrid(_ service: RealDebridService, hash: String) async throws -> StreamInfo {
        var torrentId: String

        // Step 1: Check if already in user's torrents
        if let existingId = try? await service.findExistingTorrent(hash: hash) {
            #if DEBUG
            print("[DebridManager] Found existing torrent: \(existingId)")
            #endif
            torrentId = existingId
        } else {
            // Step 2: Add magnet (with retry logic built into RealDebridService)
            torrentId = try await service.addMagnet(hash: hash)
            #if DEBUG
            print("[DebridManager] Added magnet, torrent ID: \(torrentId)")
            #endif
        }

        // Step 3-5: Select files and get stream URL (with polling)
        return try await service.getStreamURL(torrentId: torrentId)
    }

    /// Validate all configured services and return status.
    ///
    /// Services are validated concurrently (mirroring `checkCacheAll`) instead of
    /// in a serial loop, so total latency is bounded by the slowest service rather
    /// than their sum. Each task carries its `services` index so the returned array
    /// preserves the original service order regardless of completion order — the
    /// return shape/semantics are unchanged from the serial version.
    func validateAll() async -> [(DebridServiceType, Bool)] {
        var collected: [(index: Int, serviceType: DebridServiceType, valid: Bool)] = []
        await withTaskGroup(of: (Int, DebridServiceType, Bool).self) { group in
            for (index, service) in services.enumerated() {
                group.addTask {
                    let valid = (try? await service.validateToken()) ?? false
                    return (index, service.serviceType, valid)
                }
            }
            for await result in group {
                collected.append((result.0, result.1, result.2))
            }
        }
        return collected
            .sorted { $0.index < $1.index }
            .map { ($0.serviceType, $0.valid) }
    }
}

File: web/src/services/debrid/DebridManager.ts
// Port of Sources/DebridStreamer/Services/Debrid/DebridManager.swift.
//
// Manages multiple debrid services and routes requests to the best available one.
// Mirrors the Swift actor's order-independent, priority-driven merge in
// checkCacheAll, the resolveStream preferred/priority selection (with the
// Real-Debrid smart findExistingTorrent flow), and the concurrent, order-
// preserving validateAll. The keychain/secret-resolution `configure(configs:)`
// path is intentionally not ported here (no DB/keychain in the web layer);
// services are registered directly via addService, as the Swift tests do.

import { CacheStatus, type DebridServiceType, DebridServiceType as DebridServiceTypeNS, type DebridTorrent, type EpisodeFileHint, type StreamInfo } from "./models";
import { type DebridService, DebridError } from "./types";
import { RealDebridService } from "./RealDebridService";

/** Per-hash merged cache result: the bound service plus its reported status. */
export interface MergedCacheEntry {
  service: DebridServiceType;
  status: CacheStatus;
}

/** Maximum time a single provider may hold up a cache lookup. A provider that
 * exceeds this budget contributes no answer, while healthy providers continue
 * to determine the result. */
export const DEBRID_CACHE_TIMEOUT_MS = 10_000;

function withTimeout<T>(
  promise: Promise<T>,
  ms: number,
  label: string,
  signal?: AbortSignal,
): Promise<T> {
  if (signal?.aborted) {
    return Promise.reject(new Error(`${label} cache check aborted`));
  }

  return new Promise<T>((resolve, reject) => {
    let settled = false;
    const finish = (callback: () => void) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      signal?.removeEventListener("abort", onAbort);
      callback();
    };
    const onAbort = () => {
      finish(() => reject(new Error(`${label} cache check aborted`)));
    };
    const timer = setTimeout(() => {
      finish(() => reject(new Error(`${label} cache check timed out after ${ms}ms`)));
    }, ms);

    signal?.addEventListener("abort", onAbort, { once: true });
    promise.then(
      (value) => finish(() => resolve(value)),
      (error) => finish(() => reject(error)),
    );
  });
}

export class DebridManager {
  private services: DebridService[] = [];

  /** Add a single service. Mirrors Swift `addService`. Insertion order is the
   * priority order (services[0] is highest priority). */
  addService(service: DebridService): void {
    this.services.push(service);
  }

  /** All active service types. Mirrors `activeServiceTypes`. */
  get activeServiceTypes(): DebridServiceType[] {
    return this.services.map((s) => s.serviceType);
  }

  /** Mirrors `hasServices`. */
  get hasServices(): boolean {
    return this.services.length > 0;
  }

  /** Check cache across all active services, binding each hash to the highest-
   * priority service that reports it (cached beats non-cached; ties broken by the
   * lower priority index). The merge is deterministic regardless of completion
   * order. Mirrors Swift `checkCacheAll`. */
  async checkCacheAll(
    hashes: string[],
    signal?: AbortSignal,
  ): Promise<Record<string, MergedCacheEntry>> {
    if (hashes.length === 0) return {};

    // Query all services concurrently, carrying each service's priority index.
    const collected = await Promise.all(
      this.services.map(async (service, index) => {
        let cache: Record<string, CacheStatus> = {};
        try {
          cache = await withTimeout(
            service.checkCache(hashes),
            DEBRID_CACHE_TIMEOUT_MS,
            DebridServiceTypeNS.displayName(service.serviceType),
            signal,
          );
        } catch {
          cache = {};
        }
        return { index, serviceType: service.serviceType, cache };
      }),
    );

    // Deterministic merge by priority index.
    const results: Record<
      string,
      { index: number; service: DebridServiceType; status: CacheStatus }
    > = {};
    for (const entry of collected.sort((a, b) => a.index - b.index)) {
      for (const [rawHash, status] of Object.entries(entry.cache)) {
        // Torrent infoHashes are case-insensitive hex, but providers echo them
        // in different cases (Real-Debrid tends uppercase, others lowercase).
        // Canonicalize to lowercase so (a) the caller's lookup can't miss on a
        // case mismatch - which made cached streams read as uncached - and
        // (b) the same torrent from two providers merges instead of splitting.
        const hash = rawHash.toLowerCase();
        const existing = results[hash];
        if (existing) {
          const beatsCacheState =
            CacheStatus.isCached(status) && !CacheStatus.isCached(existing.status);
          const sameCacheStateLowerIndex =
            CacheStatus.isCached(status) === CacheStatus.isCached(existing.status) &&
            entry.index < existing.index;
          if (beatsCacheState || sameCacheStateLowerIndex) {
            results[hash] = {
              index: entry.index,
              service: entry.serviceType,
              status,
            };
          }
        } else {
          results[hash] = {
            index: entry.index,
            service: entry.serviceType,
            status,
          };
        }
      }
    }

    const out: Record<string, MergedCacheEntry> = {};
    for (const [hash, v] of Object.entries(results)) {
      out[hash] = { service: v.service, status: v.status };
    }
    return out;
  }

  /** Resolve a torrent hash to a stream URL using the preferred (or first
   * available) service. Mirrors Swift `resolveStream`. No silent cross-service
   * fallback: the selected service's error propagates. */
  async resolveStream(
    hash: string,
    preferredService: DebridServiceType | null = null,
    fileHint: EpisodeFileHint | null = null,
  ): Promise<StreamInfo> {
    const service = this.pickService(preferredService);

    // For Real-Debrid, use the smart resolve flow.
    if (service instanceof RealDebridService) {
      return this.resolveWithRealDebrid(service, hash, fileHint);
    }

    // Generic flow for other services.
    const torrentId = await service.addMagnet(hash);
    await service.selectFiles(torrentId, []);
    return service.getStreamURL(torrentId, fileHint);
  }

  /** Smart Real-Debrid resolve flow: reuse an existing torrent if present, else
   * add the magnet, then poll for the stream URL. Mirrors `resolveWithRealDebrid`. */
  private async resolveWithRealDebrid(
    service: RealDebridService,
    hash: string,
    fileHint: EpisodeFileHint | null = null,
  ): Promise<StreamInfo> {
    let torrentId: string;

    let existingId: string | null = null;
    try {
      existingId = await service.findExistingTorrent(hash);
    } catch {
      existingId = null; // mirrors Swift `try?`
    }

    if (existingId != null) {
      torrentId = existingId;
    } else {
      torrentId = await service.addMagnet(hash);
    }

    return service.getStreamURL(torrentId, fileHint);
  }

  /** Get a Real-Debrid transcoded HLS (`.m3u8`) URL for an already-resolved
   * stream, so a non-browser-playable container (MKV/HEVC/AV1) can be played
   * in-webview via hls.js instead of being handed off to a native player.
   *
   * Returns null (the caller falls back to the native-player path) when:
   *   - the stream wasn't resolved by Real-Debrid (no `restrictedId`, or RD
   *     isn't a configured service), or
   *   - Real-Debrid has no HLS transcode available for it.
   *
   * Never throws for the "not available" case; a network/HTTP failure from the
   * transcode call is swallowed to null so playback degrades gracefully. */
  async getTranscodeHLS(stream: StreamInfo): Promise<string | null> {
    const id = stream.restrictedId;
    if (id == null || id.length === 0) return null;
    const rd = this.services.find(
      (s): s is RealDebridService => s instanceof RealDebridService,
    );
    if (rd == null) return null;
    try {
      return await rd.getTranscodeHLS(id);
    } catch {
      return null;
    }
  }

  /** Add a magnet (by infoHash) to a debrid account so it gets cached there.
   * Uses the preferred service when given, else the highest-priority one.
   * Returns the service-native torrent/transfer id. Surfaced for the hash-list
   * import flow (bulk-hydrate a list of hashes onto the user's debrid). */
  async addMagnet(
    hash: string,
    preferredService: DebridServiceType | null = null,
  ): Promise<string> {
    const service = this.pickService(preferredService);
    return service.addMagnet(hash);
  }

  /** List the account's torrents across every service that supports listing,
   * concurrently and fault-tolerantly (a service that throws contributes no
   * rows rather than failing the whole call). Service order is preserved, with
   * each service's rows kept together. The Debrid Library manager's data source. */
  async listTorrents(): Promise<DebridTorrent[]> {
    const collected = await Promise.all(
      this.services.map(async (service, index) => {
        if (typeof service.listTorrents !== "function") {
          return { index, rows: [] as DebridTorrent[] };
        }
        try {
          return { index, rows: await service.listTorrents() };
        } catch {
          return { index, rows: [] as DebridTorrent[] };
        }
      }),
    );
    return collected
      .sort((a, b) => a.index - b.index)
      .flatMap((c) => c.rows);
  }

  /** Delete a torrent/transfer by id from a specific service (matched by its
   * short code, e.g. "RD"/"AD"). Throws if no matching service supports delete. */
  async deleteTorrent(id: string, debridServiceCode: string): Promise<void> {
    for (const service of this.services) {
      if (
        DebridServiceTypeNS.shortCode(service.serviceType) === debridServiceCode &&
        typeof service.deleteTorrent === "function"
      ) {
        await service.deleteTorrent(id);
        return;
      }
    }
    throw DebridError.networkError(
      `No configured service can delete torrents for ${debridServiceCode}`,
    );
  }

  /** Resolve a single service by preference, else the highest priority one. */
  private pickService(preferredService: DebridServiceType | null): DebridService {
    if (preferredService != null) {
      const found = this.services.find((s) => s.serviceType === preferredService);
      if (found == null) {
        throw DebridError.networkError(
          `Service ${DebridServiceTypeNS.displayName(preferredService)} not configured`,
        );
      }
      return found;
    }
    const first = this.services[0];
    if (first == null) {
      throw DebridError.networkError("No debrid services configured");
    }
    return first;
  }

  /** Validate all configured services concurrently, preserving service order in
   * the result. Mirrors Swift `validateAll`. */
  async validateAll(): Promise<[DebridServiceType, boolean][]> {
    const collected = await Promise.all(
      this.services.map(async (service, index) => {
        let valid = false;
        try {
          valid = await service.validateToken();
        } catch {
          valid = false;
        }
        return { index, serviceType: service.serviceType, valid };
      }),
    );
    return collected
      .sort((a, b) => a.index - b.index)
      .map((c) => [c.serviceType, c.valid] as [DebridServiceType, boolean]);
  }
}
